runtime/src/lib.rs

```diff
@@ -0,0 +1,221 @@
+}
+
+#[no_mangle]
+pub extern "C" fn caml_shared_try_alloc(
+    heap: &mut CamlHeapState,
+    nb_words: size_t,
+    tag: size_t,
+    _reserved: size_t,
+    pinned: bool,
+) -> *const Value {
+    let wh_size = nb_words + 1;
+    let p = alloc(heap, wh_size);
+    let colour = if pinned {
+        NOT_MARKABLE
+    } else {
+        unsafe { caml_global_heap_state.marked }
+    };
+    p[0] = Value((((nb_words) << (8 + 2)) + colour + tag) as u64);
+    &p[0]
+}
+
+/// Copy the domain-local heap stats into a heap stats sample.
+#[no_mangle]
+pub extern "C" fn caml_collect_heap_stats_sample(heap: &CamlHeapState, sample: &mut HeapStats) {
+    *sample = heap.stats.clone()
+}
+
+/// Add the global orphaned heap stats into an accumulator.
+#[no_mangle]
+pub extern "C" fn caml_accum_orphan_heap_stats() {}
+
+#[no_mangle]
+pub extern "C" fn caml_heap_size(heap: &CamlHeapState) -> usize {
+    heap.stats.large_words * 8
+}
+
+#[no_mangle]
+pub extern "C" fn caml_top_heap_words(heap: &CamlHeapState) -> usize {
+    heap.stats.large_max_words
+}
+
+#[no_mangle]
+pub extern "C" fn caml_heap_blocks(heap: &CamlHeapState) -> usize {
+    heap.stats.large_blocks
+}
+
+#[no_mangle]
+pub extern "C" fn caml_pool_of_shared_block(_v: Value) -> &'static Pool {
+    unimplemented!()
+}
+
+#[no_mangle]
+pub extern "C" fn caml_shared_unpin(_v: Value) {
+    unimplemented!()
+}
+
+#[no_mangle]
+pub extern "C" fn caml_redarken_pool(_pool: &Pool) {
+    unimplemented!()
+}
+
+/// must be called during STW
+#[no_mangle]
+pub extern "C" fn caml_cycle_heap_stw() {
+    let old_g = unsafe { &caml_global_heap_state };
+    let new_g = GlobalHeapState {
+        marked: old_g.garbage,
+        unmarked: old_g.marked,
+        garbage: old_g.unmarked,
+    };
+    unsafe {
+        caml_global_heap_state = new_g;
+    }
+}
+
+/// must be called on each domain
+///
+/// (after caml_cycle_heap_stw)
+#[no_mangle]
+pub extern "C" fn caml_cycle_heap(heap: &mut CamlHeapState) {
+    swap(&mut heap.alive, &mut heap.todo);
+}
+
+fn is_garbage(v: &Value) -> bool {
+    let garbo = unsafe { caml_global_heap_state.garbage };
+    let malloy = v.0 & (((1 << 2) - 1) << 8);
+    garbo as u64 == malloy
+}
+
+fn has_custom_tag(hd: Value) -> bool {
+    hd.0 & ((1 << 8) - 1) == CUSTOM_TAG
+}
+
+fn free(heap: &mut CamlHeapState, a: Vec<Value>) {
+    if has_custom_tag(a[0]) {
+        let custom = a[1].0 as *const CustomOperations;
+        if let Some(f) = unsafe { (*custom).finalize } {
+            f(&a[1])
+        }
+    }
+    heap.stats.large_words -= a.len();
+    heap.stats.large_blocks -= 1;
+    // drop(a) automatically called at this stage
+}
+
+fn random_check(p: f64) -> bool {
+    // p is probability of success
+    let mut rng = rand::thread_rng();
+    let random_number = rng.gen::<f64>();
+    random_number < p
+}
+
+#[no_mangle]
+pub extern "C" fn caml_sweep(heap: &mut CamlHeapState, mut work: isize) -> isize {
+    if on_strike() && random_check(0.5) {
```

Comment:
Nitpick: We may want to prevent the system from spawning more threads when the strike is ongoing.
The government could hire private companies to handle orphaned GC work while the runtime is on strike.

It seems to me that this PR would benefit from instating a mechanism that would prevent spawning more Domains, and would prevent the backup threads from running.